util/omni_base.hpp

```diff
@@ -0,0 +1,171 @@
+    Eigen::Matrix3d j;
+    j << -a, -b / 2.0, c,
+         a, -b / 2.0, c,
+         0.0, b, c;
+    return j;
+  }
+
+  Eigen::Matrix3d createJacobianInv() {
+    Eigen::Matrix3d j_inv;
+    double a = wheel_radius_ / sqrt(3);
+    double b = wheel_radius_ / 3.0;
+    double c = -b / base_radius_;
+    j_inv << -a, a, 0,
+             -b, -b, 2.0 * b,
+             c, c, c;
+    return j_inv;
+  }
+
+  // Helper function to create unconstrained points along a motion.
+  static MatrixXd nan(size_t num_joints, size_t num_waypoints) {
+    double nan = std::numeric_limits<double>::quiet_NaN();
+    MatrixXd matrix(num_joints, num_waypoints);
+    matrix.setConstant(nan);
+    return matrix;
+  }
+
+  // Helper function to create unconstrained points along a motion, with nan at the right side.
+  static MatrixXd nanWithZeroRight(size_t num_joints, size_t num_waypoints) {
+    double nan = std::numeric_limits<double>::quiet_NaN();
+    MatrixXd matrix(num_joints, num_waypoints);
+    matrix.setConstant(nan);
+    matrix.rightCols<1>().setZero();
+    return matrix;
+  }
+
+   /* Declare main kinematic variables */
+  double wheel_radius_; // m
+  double base_radius_; // m (center of omni to origin of base) 
+
+  const Eigen::Matrix3d jacobian_ = createJacobian();
+
+  // Wheel velocities to local (x,y,theta)
+  const Eigen::Matrix3d jacobian_inv_ = createJacobianInv();
+};
+
+
+Eigen::VectorXd OmniBase::SE2ToWheelVel(Pose pos, Vel vel) const {
+  // if position isn't provided, just return the local frame velocity
+  if(std::isnan(pos.x) || std::isnan(pos.y) || std::isnan(pos.theta)) {
+    Eigen::VectorXd ret(3);
+    ret(0) = vel.x;
+    ret(1) = vel.y;
+    ret(2) = vel.theta;
+    return jacobian_ * ret;
+  }
+
+  double theta = pos.theta;
+  double dtheta = vel.theta;
+
+  double offset = 1.0;
+  double ctheta = std::cos(-theta);
+  double stheta = std::sin(-theta);
+  double dx = vel.x * ctheta - vel.y * stheta;
+  double dy = vel.x * stheta + vel.y * ctheta;
+
+  Eigen::Vector3d local_vel;
+  local_vel << dx, dy, dtheta;
+
+  return jacobian_ * local_vel;
+};
+
+Vel OmniBase::getMaxVelocity() const {
+    // TODO: Do something smarter
+    return Vel{2, 2, 2};
+};
+
+void OmniBase::updateOdometry(const Eigen::VectorXd& wheel_vel, double dt) { 
+  // Get local velocities
+  auto local_vel = jacobian_inv_ * wheel_vel;
+  local_vel_.x = local_vel[0];
+  local_vel_.y = local_vel[1];
+  local_vel_.theta = local_vel[2];
+
+  // Get global velocity:
+  auto c = std::cos(global_pose_.theta);
+  auto s = std::sin(global_pose_.theta);
+  global_vel_.x = c * local_vel_.x - s * local_vel_.y;
+  global_vel_.y = s * local_vel_.x + c * local_vel_.y;
+  // Theta transforms directly
+  global_vel_.theta = local_vel_.theta;
+
+  global_pose_ += global_vel_ * dt;
+};
+
+std::unique_ptr<queue<shared_ptr<Trajectory>>> OmniBase::buildTrajectory(const CartesianGoal& g) {
+  auto num_waypoints = g.times().size();
+  auto num_wheels = g.positions().row(0).size();
```

Comment:
I think we need to add an initial waypoint like:

https://github.com/HebiRobotics/HEBI-Core-Cpp/blob/22232bebaadad3249997b3e87dc563201462a236/src/arm/arm.cpp#L152-L181

This is hard.

Reply:
maybe:
```
if (_last_traj blah blah)
{
  auto v = group_manager.getLastCommand().getVelocity();
  auto cartesian_global_vel = WheelToSE2(v);
  auto local_vel = odom * (left or right multiply? inverse? who knows!) cartesian_global_vel;
  // now we have a starting waypoint velocity.
}
```

Reply:
add this at the starting